roachtest: sqlsmith/setup=seed/setting=no-mutations failed

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913125&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2913125&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [5f40d691ef9caf910f97fbad53802000c0ca72eb](https://github.com/cockroachdb/cockroach/commits/5f40d691ef9caf910f97fbad53802000c0ca72eb):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:248,sqlsmith.go:262,test_runner.go:777: ping: dial tcp 34.72.246.207:26257: connect: connection refused
		previous sql:
		SELECT
			string_agg(tab_1079._bytes::BYTES, tab_1079._bytes::BYTES) OVER (PARTITION BY tab_1079._bytes ORDER BY tab_1079._date RANGE CURRENT ROW)::BYTES
				AS col_2759,
			tab_1079._date AS col_2760
		FROM
			defaultdb.public.seed@seed__int8__float8__date_idx AS tab_1079
		WHERE
			tab_1079._bool
		GROUP BY
			tab_1079._bytes, tab_1079._date
		LIMIT
			35:::INT8;

	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2913125-1619118901-10-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: dead
		3: 13739
		4: 13417
		2: 14808
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1162
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:284
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2057
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=seed/setting=no-mutations
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xeee982]

goroutine 2719 [running]:
panic(0x48c3720, 0xc0033a33b0)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc004716ff8 sp=0xc004716f30 pc=0x48a945
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc000344600, 0x5bcc3a0, 0xc0006d86c0, 0x466d8e0, 0x7ecd190)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:814 +0x3a9 fp=0xc004717078 sp=0xc004716ff8 pc=0x3599ea9
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc000344600, 0x5bcc3a0, 0xc0006d86c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482 +0x65 fp=0xc0047170b0 sp=0xc004717078 pc=0x376e3c5
runtime.call32(0x0, 0x544b720, 0xc000defe48, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc0047170e0 sp=0xc0047170b0 pc=0x4c22de
runtime.reflectcallSave(0xc004717220, 0x544b720, 0xc000defe48, 0x18)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc004717110 sp=0xc0047170e0 pc=0x48a338
runtime.runOpenDeferFrame(0xc002176000, 0xc000defe00, 0x0)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc0047171a0 sp=0xc004717110 pc=0x48a1ed
panic(0x466d8e0, 0x7ecd190)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc004717268 sp=0xc0047171a0 pc=0x48a5b9
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc004717298 sp=0xc004717268 pc=0x4a1693
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0x4c0d1b)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc0047172d0 sp=0xc004717298 pc=0xeee982
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog(0x0, 0x0, 0x3, 0xc000101000, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:94 +0x39 fp=0xc004717308 sp=0xc0047172d0 pc=0x1017ef9
github.com/cockroachdb/cockroach/pkg/util/log.vEventf(0x0, 0x0, 0x462d100, 0x1, 0xc000000001, 0x4d988b0, 0x19, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:242 +0x12a fp=0xc004717438 sp=0xc004717308 pc=0x10192ea
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:263
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*externalSorter).Close(0xc0042b81a0, 0x0, 0x0, 0x4d42fe1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_sort.go:534 +0x8c fp=0xc0047174a8 sp=0xc004717438 pc=0x2d628cc
github.com/cockroachdb/cockroach/pkg/sql/colexecop.Closers.CloseAndLogOnErr(0xc002c04990, 0x3, 0x9, 0x0, 0x0, 0xc003164190, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:160 +0xe6 fp=0xc004717540 sp=0xc0047174a8 pc=0x29cd506
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).close(0xc003f00600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:333 +0x9e fp=0xc004717588 sp=0xc004717540 pc=0x2d6c61e
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).ConsumerClosed(0xc003f00600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:339 +0x2b fp=0xc0047175a0 sp=0xc004717588 pc=0x2d6c68b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).InternalClose(0xc00489e000, 0xc003164180)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:951 +0xae fp=0xc0047175e0 sp=0xc0047175a0 pc=0x246304e
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).close(0xc00489e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:243 +0x31 fp=0xc004717628 sp=0xc0047175e0 pc=0x272aa31
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*windower).ConsumerClosed(0xc00489e000)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/windower.go:239 +0x2b fp=0xc004717640 sp=0xc004717628 pc=0x272a9eb
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).InternalClose(0xc00268c000, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:951 +0xae fp=0xc004717680 sp=0xc004717640 pc=0x246304e
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Close(0xc00268c000, 0x5bcc3a0, 0xc00429c480, 0x4d42fe1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:307 +0x46 fp=0xc0047176a0 sp=0xc004717680 pc=0x2d5ec06
github.com/cockroachdb/cockroach/pkg/sql/colexecop.Closers.CloseAndLogOnErr(0xc004248d00, 0x1, 0xd, 0x5bcc3a0, 0xc00429c480, 0xc003164180, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:160 +0xe6 fp=0xc004717738 sp=0xc0047176a0 pc=0x29cd506
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).close(0xc00340ec00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:333 +0x9e fp=0xc004717780 sp=0xc004717738 pc=0x2d6c61e
github.com/cockroachdb/cockroach/pkg/sql/colexec.NewMaterializer.func1(0x4b533a0, 0xc0015851f0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x2a fp=0xc004717798 sp=0xc004717780 pc=0x2dd9eaa
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).moveToTrailingMeta(0xc00340ec00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:755 +0x24e fp=0xc0047178d8 sp=0xc004717798 pc=0x2461cae
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper(0xc00340ec00, 0x2463193)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:656 +0x19d fp=0xc004717950 sp=0xc0047178d8 pc=0x246187d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next(0xc00340ec00, 0xc002bdffc0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:325 +0xe5 fp=0xc004717988 sp=0xc004717950 pc=0x2d6c545
github.com/cockroachdb/cockroach/pkg/sql/execinfra.DrainAndForwardMetadata(0x5bea860, 0xc0042b84b8, 0x5bf1220, 0xc00340ec00, 0x5b94920, 0xc003c05880)
```

--- a/pkg/sql/colexec/materializer.go
+++ b/pkg/sql/colexec/materializer.go
@@ -332,6 +332,14 @@ func (m *Materializer) close() {
 		if m.cancelFlow != nil {
 			m.cancelFlow()()
 		}
+		if m.Ctx == nil {
+			// In some edge cases (like when Init of an operator above this
+			// materializer encounters a panic), the materializer might never be
+			// started, yet it still will attempt to close its Closers. This
+			// context is only used for logging purposes, so it is ok to grab
+			// the background context in order to prevent a NPE below.
+			m.Ctx = context.Background()
+		}
 		m.closers.CloseAndLogOnErr(m.Ctx, "materializer")
 	}
 }

--- a/pkg/sql/colflow/panic_injector.go
+++ b/pkg/sql/colflow/panic_injector.go
@@ -17,6 +17,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecop"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/errors"
 )
@@ -25,6 +26,7 @@ import (
 // Init and Next methods of the wrapped operator.
 type panicInjector struct {
 	colexecop.OneInputNode
+	colexecop.InitHelper
 	rng *rand.Rand
 }
 
@@ -50,14 +52,19 @@ func newPanicInjector(input colexecop.Operator) colexecop.Operator {
 }
 
 func (i *panicInjector) Init(ctx context.Context) {
+	if !i.InitHelper.Init(ctx) {
+		return
+	}
 	if i.rng.Float64() < initPanicInjectionProbability {
+		log.Info(i.Ctx, "injecting panic in Init")
 		colexecerror.ExpectedError(errors.New("injected panic in Init"))
 	}
-	i.Input.Init(ctx)
+	i.Input.Init(i.Ctx)
 }
 
 func (i *panicInjector) Next() coldata.Batch {
 	if i.rng.Float64() < nextPanicInjectionProbability {
+		log.Info(i.Ctx, "injecting panic in Next")
 		colexecerror.ExpectedError(errors.New("injected panic in Next"))
 	}
 	return i.Input.Next()

